### PR TITLE
Preview PR: implemented rich man's pagination via temp table

### PR DESCRIPTION
**Pull request by dvoet**
This is a preview of the rich man. I would like feedback on the overall approach and a formal PR will come later.

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala [hunk 1]
```diff
@@ -48,6 +49,13 @@ trait EntityComponent {
       }
     }
 
+    private object PaginatedEntityRawSqlQuery extends RawSqlQuery {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala:52]: ignore this

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala [hunk 2]
```diff
@@ -109,6 +117,90 @@ trait EntityComponent {
...
+
+        def order(alias: String) = sql" order by #$alias.sort_field #${SortDirections.toSql(entityQuery.sortDirection)} "
+
+        val paginationJoin = concatSqlActions(
+          sql""" join (select pagination.id, pagination.name, pagination.sort_field from ENTITY_PAGINATION pagination """,
+          filterSql,
+          order("pagination"),
+          sql" limit #${entityQuery.pageSize} offset #${(entityQuery.page-1) * entityQuery.pageSize} ) p on p.id = e.id "
+        )
+
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala:184]: the following for loop outlines the general approach:
- populate a temp table with all the entities, all the attributes of an entity crammed into 1 column for filtering, and a column with the thing that should be sorted
- do the normal enitity query but join that to a sub select on the temp table with the appropriate filter, sort, limit and offset

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala [hunk 3]
```diff
@@ -169,21 +261,24 @@ trait EntityComponent {
       unmarshalEntitiesWithIds(entityAttributeAction).map(_.map { case (id, entity) => entity })
     }
 
-    def unmarshalEntitiesWithIds(entityAttributeAction: ReadAction[Seq[EntityAndAttributesRawSqlQuery.EntityAndAttributesResult]]): ReadAction[Map[Long, Entity]] = {
-      entityAttributeAction.map { entityAttributeRecords =>
-        val allEntityRecords = entityAttributeRecords.map(_.entityRecord).toSet
+    def unmarshalEntitiesWithIds(entityAttributeAction: ReadAction[Seq[EntityAndAttributesRawSqlQuery.EntityAndAttributesResult]]): ReadAction[mutable.LinkedHashMap[Long, Entity]] = {
+      entityAttributeAction.map(unmarshalEntitiesWithIds)
+    }
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala:267]: these changes to use LinkedHashMap are to preserve ordering from the sql results

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiServiceSpec.scala [hunk 4]
```diff
@@ -590,284 +590,285 @@ class EntityApiServiceSpec extends ApiServiceSpec {
   val defaultQuery = EntityQuery(1, 10, "name", Ascending, None)
   def calculateNumPages(count: Int, pageSize: Int) = Math.ceil(count.toDouble / pageSize).toInt
```
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiServiceSpec.scala:592]: these test changes are to be able to run the same tests for both rich and poor

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 5]
```diff
@@ -639,8 +650,8 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
         (x, y) match {
           case (AttributeNumber(a), AttributeNumber(b)) => a.compare(b)
           case (AttributeNull, AttributeNull) => 0
-          case (AttributeNull, _) => 1
-          case (_, AttributeNull) => -1
+          case (AttributeNull, _) => -1
+          case (_, AttributeNull) => 1
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:654]: the database sorts null differently

File: src/main/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiService.scala [hunk 6]
```diff
@@ -39,8 +39,13 @@ trait EntityApiService extends HttpService with PerRequestCreator with UserInfoD
 
           if (errors.isEmpty) {
             val entityQuery = EntityQuery(toIntTries("page").get.getOrElse(1), toIntTries("pageSize").get.getOrElse(10), sortField.getOrElse("name"), sortDirectionTry.get, filterTerms)
-            requestContext => perRequest(requestContext, WorkspaceService.props(workspaceServiceConstructor, userInfo),
-              WorkspaceService.QueryEntities(WorkspaceName(workspaceNamespace, workspaceName), entityType, entityQuery))
+            requestContext =>
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/webservice/EntityApiService.scala:42]: support both rich and poor simultaneously (for now)

**Comment by dmohs** [comment]
Overall I'm a huge fan. One question: Would it be relatively straightforward to duplicate the "sort column" stuff so that it would be possible in the future for a user to sort by X then Y?

**Pull request closed without merging**